### PR TITLE
fix(pkg/repository/maintenance): handle when there's no container status

### PR DESCRIPTION
**Pull request by mcluseau**
Fixes a crash in this case.

```
panic: runtime error: index out of range [0] with length 0 [recovered]
	panic: runtime error: index out of range [0] with length 0

goroutine 562 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:116 +0x1e5
panic({0x2ae2cc0?, 0xc001a084b0?})
	/usr/local/go/src/runtime/panic.go:770 +0x132
github.com/vmware-tanzu/velero/pkg/repository.GetMaintenanceResultFromJob({0x318cd00, 0xc0009be3f0}, 0xc0025ae508)
	/go/src/github.com/vmware-tanzu/velero/pkg/repository/maintenance.go:120 +0x229
```

### CODE REVIEW

File: pkg/repository/maintenance.go
Comment (mateusoliveira43) [inline pkg/repository/maintenance.go:122]: Since function wants result from job, what about changing error message to something like
```
container statuses not yet populated for <job>
```
?

**Review by mateusoliveira43 — APPROVED** [review]

**Comment by mateusoliveira43** [comment]
@mcluseau  yeah, Go errors are almost unreadable, thanks for taking time to fix this!

**Review by mateusoliveira43 — APPROVED** [review]

**Review by kaovilai — APPROVED** [review]

**Review by shubham-pampattiwar — DISMISSED** [review]

**Comment by kaovilai** [comment]
@mcluseau please add changelog

**Comment by kaovilai** [comment]
Add changelog here
https://github.com/vmware-tanzu/velero/tree/main/changelogs/unreleased
File name: `<pr-num>-<gh-username>`

File: changelogs/unreleased/8271-mcluseau
Comment (kaovilai) [inline changelogs/unreleased/8271-mcluseau:1]:
```suggestion
fix(pkg/repository/maintenance): don't panic when there's no container statuses
```

Comment (mcluseau) [reply changelogs/unreleased/8271-mcluseau]: yup, I hope it's not too bold :'D

File: pkg/repository/maintenance.go
Comment (Lyndon-Li) [inline pkg/repository/maintenance.go:124]: The termination-log is not always written, it is only written when maintenance fails.
So I suspect it is a non-error case when `len(statuses) == 0`

File: pkg/repository/maintenance.go
Comment (mcluseau) [reply pkg/repository/maintenance.go:122]: I don't have the job's pod anymore, sorry.
I didn't have any issue since I applied my fix, but I agree it's better to be even safer here.

**Review by kaovilai — APPROVED** [review]

**Review by shubham-pampattiwar — APPROVED** [review]

File: pkg/repository/maintenance.go
Comment (Lyndon-Li) [reply pkg/repository/maintenance.go:124]: Sorry I may have pointed the comment to wrong code. The most suspecting line is below:
```
	terminated := statuses[0].State.Terminated
	if terminated == nil {
		return "", fmt.Errorf("container for job %s is not terminated", job.Name)
```
As you can see from the [maintenance code](https://github.com/vmware-tanzu/velero/blob/f02613d2f7a7cfbf687f84d1deac7414d679569e/pkg/cmd/cli/repomantenance/maintenance.go#L71), if the maintenance succeeds, no termination message is written, so `statuses[0].State.Terminated` may indeed be nil. But this is a normal case.

Furthermore, when nothing is set to `pod.Status.ContainerStatuses `, will it be nil? If so, it is still a normal case.

Please double confirm this.

Comment (mcluseau) [reply pkg/repository/maintenance.go:124]: I answered to the `State.Terminated` case :) there's not just the termination log, there's also the exit code, time, [etc](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.29/#containerstateterminated-v1-core).

About container statuses, they are expected to be filled (from waiting to running to terminated) but they may not be present when the pod was just created (I guess a kind of race condition leads the initial but mentioned here). So, my interpretation is that when it's empty, it's not a normal state and can't be after the container's termination: ie, the caller made the call too soon for some reason, very in line with semantics of asking before the pod was created by the job, which is a case that returns an error.

**Review by kaovilai — APPROVED** [review]

**Comment by Lyndon-Li** [comment]
@mcluseau 
Please don't resolve the comments until consensus are made with the commenter, otherwise, maintainers may think the PR is ready and merge it.

File: pkg/repository/maintenance.go
Comment (mcluseau) [reply pkg/repository/maintenance.go:124]: I was explaining why «if X, no termination message is written, so `statuses[0].State.Terminated` may indeed be nil» is not true. Container statuses are set by the kubelet, so the authority is the k8s' spec, that also explain why terminated *must* be set to hold other information about the container's termination. As you asked, I double checked this basic assumption I had by going to spec.

**Comment by mcluseau** [comment]
> @mcluseau Please don't resolve the comments until consensus are made with the commenter, otherwise, maintainers may think the PR is ready and merge it.

2 weeks looked like a lazy consensus to me, but ok, I won't touch those buttons again :sweat_smile:

File: pkg/repository/maintenance.go
Comment (mcluseau) [reply pkg/repository/maintenance.go:122]: @blackpiglet this should be solved, please kindly confirm by resolving this conversation, thanks.

File: pkg/repository/maintenance.go
Comment (Lyndon-Li) [reply pkg/repository/maintenance.go:124]: OK, I got you. `statuses[0].State.Terminated` should always be set as long as the pod reaches to one of the terminated states.

**Comment by Lyndon-Li** [comment]
We are creating release of 1.15, let's merge this PR after 1.15 branch is cut.

**Review by Lyndon-Li — APPROVED** [review]

**Review by blackpiglet — APPROVED** [review]

**Review by shubham-pampattiwar — APPROVED** [review]

**Pull request merged**
